pkg/automationconfig/automation_config_secret.go

```diff
@@ -0,0 +1,66 @@
+package automationconfig
+
+import (
+	"encoding/json"
+
+	"github.com/mongodb/mongodb-kubernetes-operator/pkg/kube/secret"
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const ConfigKey = "cluster-config.json"
+
+// EnsureSecret makes sure that the AutomationConfig secret exists with the desired config.
+// if the desired config is the same as the current contents, no change is made.
+// The most recent AutomationConfig is returned. If no change is made, it will return the existing one, if there
+// is a change, the new AutomationConfig is returned.
+func EnsureSecret(secretGetUpdateCreator secret.GetUpdateCreator, secretNsName types.NamespacedName, owner []metav1.OwnerReference, desiredAutomationConfig AutomationConfig) (AutomationConfig, error) {
+	existingSecret, err := secretGetUpdateCreator.GetSecret(secretNsName)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			return createNewAutomationConfigSecret(secretGetUpdateCreator, secretNsName, owner, desiredAutomationConfig)
+		}
+		return AutomationConfig{}, err
+	}
+
+	acBytes, err := json.Marshal(desiredAutomationConfig)
+	if err != nil {
+		return AutomationConfig{}, err
+	}
+	if existingAcBytes, ok := existingSecret.Data[ConfigKey]; !ok {
+		// the secret exists but the key is not present. We can update the secret
+		existingSecret.Data[ConfigKey] = acBytes
+	} else {
+		// the secret already exists, we should check to see if we're making any changes.
+		existingAutomationConfig, err := FromBytes(existingAcBytes)
+		if err != nil {
+			return AutomationConfig{}, err
+		}
+		// we are attempting to update with the same version, no change is required.
```

Comment:
[q] what's `version` here? Not the one you set [here](https://github.com/mongodb/mongodb-kubernetes-operator/pull/360/files#diff-d990f1832c6e156b4f6f747a6c3fc767f546cb53fb9a76684550a7d51b67cd71R247)?

Do you just mean "with the same content"?

Reply:
yes the same content, the agent will re-attempt to configure the state of the cluster if the version is different (event if contents are the same) so we just need to check for equal content. (while excluding the version field)